source/MaterialXShaderGen/ShaderGenerators/Glsl/AdskSurfaceGlsl.cpp

```diff
@@ -0,0 +1,84 @@
+#include <MaterialXShaderGen/ShaderGenerators/Glsl/AdskSurfaceGlsl.h>
+#include <MaterialXShaderGen/ShaderGenerators/Glsl/GlslShaderGenerator.h>
+
+namespace MaterialX
+{
+
+SgImplementationPtr AdskSurfaceGlsl::creator()
+{
+    return std::make_shared<AdskSurfaceGlsl>();
+}
+
+const string& AdskSurfaceGlsl::getLanguage() const
+{
+    return GlslShaderGenerator::LANGUAGE;
+}
+
+const string& AdskSurfaceGlsl::getTarget() const
+{
+    return GlslShaderGenerator::TARGET;
+}
+
+
+void AdskSurfaceGlsl::createVariables(const SgNode& /*node*/, ShaderGenerator& /*shadergen*/, Shader& shader_)
+{
+    // TODO: 
+    // The surface shader needs position and view data. We should solve this by adding some 
+    // dependency mechanism so this implementation can be set to depend on the PositionGlsl 
+    // and ViewGlsl nodes instead? This is where the MaterialX attribute "internalgeomprops" 
+    // is needed.
+    //
+    HwShader& shader = static_cast<HwShader&>(shader_);
+
+    shader.createAppData(DataType::VECTOR3, "i_position");
+
+    shader.createUniform(HwShader::GLOBAL_SCOPE, DataType::MATRIX4, "u_viewInverseMatrix");
+
+    shader.createVertexData(DataType::VECTOR3, "positionWorld");
+    shader.createVertexData(DataType::VECTOR3, "viewWorld");
```

Comment:
About the view vector: You are absolutely right, it's actually created as vertex data here which is wrong. It should be a uniform of course as it's the same for all vertices. I need to fix that.

About adding variables to specific stages: I had a previous design that was more general where you had to do that, but in the end I removed it because it felt to "verbose" to having to specify both outputs for vertex stage and inputs to pixel stage when they needed to match anyway. However for uniforms this might be a good idea to put back, so you could target them to a specific stage.

Reply:
I will create a new task for the view vector handling.

Reply:
Sounds good. I am all for not over complicating things. The view vector can be passed into the either one or both shader stages (so that they match when linking) so I don't think we need to have it stage specific.